File: python/Successful Pairs of Spells and Potions.py
from typing import List
import bisect

class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        potions.sort()
        result = []
        m = len(potions)
        
        for spell in spells:
            min_potion_strength = (success + spell - 1) // spell
            
            index = bisect.bisect_left(potions, min_potion_strength)

            result.append(m - index)
        
        return result

# Test Cases
def test_successful_pairs():
    solution = Solution()
    
    # Case 1
    spells = [10, 20, 30]
    potions = [1, 2, 3, 4, 5]
    success = 60
    assert solution.successfulPairs(spells, potions, success) == [3, 4, 5]
    
    # Case 2
    spells = [5, 1, 3]
    potions = [1, 2, 3, 4, 5]
    success = 7
    assert solution.successfulPairs(spells, potions, success) == [4, 0, 3]

File: js/Successful Pairs of Spells and Potions.js
var successfulPairs = function(spells, potions, success) {
    potions.sort((a, b) => a - b);  
    const result = [];
    const m = potions.length;

    const bisectLeft = (arr, target) => {
        let left = 0, right = arr.length;
        while (left < right) {
            let mid = Math.floor((left + right) / 2);
            if (arr[mid] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    };

    spells.forEach(spell => {
        let minPotionStrength = Math.ceil(success / spell);  

        let index = bisectLeft(potions, minPotionStrength);  

        result.push(m - index);  
    });

    return result;
};

// Test cases
console.log(successfulPairs([10, 15, 7], [5, 8, 12, 18], 100));  // OutPut: [2, 3, 1]
console.log(successfulPairs([3, 6, 2], [1, 2, 3, 4, 5], 8));  // OutPut: [3, 5, 2]